File: scripts/create_migration.py
import os
import re
import sys

MIGRATION_DIR = "internal/database/migrations"


class Migration:
    def __init__(self, name):
        name = re.sub(r"[^a-zA-Z0-9_]", "_", name)
        name = name.lower()
        self.name = name

    def get_migration_name(self):
        migration_files = self.get_migration_files()
        last_migration = migration_files[-1] if len(migration_files) > 0 else 0
        last_idx = int(last_migration.split("_")[0]) if last_migration else 0
        new_idx = str(last_idx + 1).zfill(4)
        file_name = f"{new_idx}_{self.name}"
        return file_name

    def create(self):
        migration_name = self.get_migration_name()
        files_to_create = [f"{migration_name}.up.sql"]

        # Check if the user wanted a down migration
        if len(sys.argv) > 2 and sys.argv[2] == "--down":
            files_to_create.append(f"{migration_name}.down.sql")

        for file in files_to_create:
            if os.path.exists(f"{MIGRATION_DIR}/{file}"):
                print(f"Migration {file} already exists.")
                continue
            with open(f"{MIGRATION_DIR}/{file}", "w") as f:
                f.write("-- Write your migration here")
                print(f"Created migration {file}")

    def get_migration_files(self):
        # Get all migration files from the migrations directory
        files = os.listdir(path=MIGRATION_DIR)
        files = [file for file in files if file.endswith(".up.sql")]
        files = [file.replace(".up.sql", "") for file in files]
        files = sorted(files)
        return files if files else []


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python create_migration.py <migration_name>")
        sys.exit(1)
    migration_name = sys.argv[1]
    migration = Migration(migration_name)
    migration.create()

File: scripts/create_migration.lua
#!/usr/bin/env lua

-- Usage:
--   ./scripts/create_migration.lua "add_users"
--   ./scripts/create_migration.lua --down "drop_users"

local dir_migrations = "migrations"

local create_down_migration = false
local migration_name = ""

-- Try to load LuaFileSystem; fall back to io.popen if unavailable
local has_lfs, lfs = pcall(require, "lfs")

-- --- helpers ---------------------------------------------------------------

local function log_info(...)
	io.stderr:write(table.concat({ ... }, " ") .. "\n")
end

local function fatal(msg)
	io.stderr:write(msg .. "\n")
	os.exit(1)
end

local function usage()
	print("Usage: ./scripts/create_migration.lua [--down] <migration_name>")
end

local function ensure_dir(path)
	if has_lfs then
		local attr = lfs.attributes(path)
		if not attr then
			local ok, err = lfs.mkdir(path)
			if not ok then
				fatal("Failed to create directory '" .. path .. "': " .. tostring(err))
			end
		elseif attr.mode ~= "directory" then
			fatal("'" .. path .. "' exists and is not a directory")
		end
	else
		os.execute(string.format("mkdir -p %q", path))
	end
end

local function list_files(path)
	local files = {}
	if has_lfs then
		for name in lfs.dir(path) do
			if name ~= "." and name ~= ".." then
				table.insert(files, name)
			end
		end
	else
		local p = io.popen("ls -1 " .. string.format("%q", path) .. " 2>/dev/null")
		if not p then
			return files
		end
		for line in p:lines() do
			table.insert(files, line)
		end
		p:close()
	end
	return files
end

local function write_file(path, contents)
	local f, err = io.open(path, "w")
	if not f then
		fatal("Error creating file '" .. path .. "': " .. tostring(err))
	end
	f:write(contents)
	f:close()
end

local function sanitize_name(name)
	name = string.lower(name or "")
	name = name:gsub("[^%w_]", "_")
	name = name:gsub("_+", "_")
	name = name:gsub("^_+", ""):gsub("_+$", "")
	return name
end

local function get_last_migration_index()
	local files = list_files(dir_migrations)
	local last = -1
	for _, fname in ipairs(files) do
		if not fname:match("/$") and fname:match("%.sql$") then
			local underscore_at = fname:find("_", 1, true)
			if underscore_at then
				local num_str = fname:sub(1, underscore_at - 1)
				local num = tonumber(num_str)
				if num and num > last then
					last = num
				end
			end
		end
	end
	return last
end

local function pad6(n)
	local s = tostring(n)
	return string.rep("0", math.max(0, 6 - #s)) .. s
end

local function create_migration()
	migration_name = sanitize_name(migration_name)
	if migration_name == "" then
		fatal("Migration name is required")
	end

	print("Creating new migration...")

	ensure_dir(dir_migrations)

	local last = get_last_migration_index()
	local new_index = (last or -1) + 1
	local index_str = pad6(new_index)

	local up_filename = string.format("%s_%s.up.sql", index_str, migration_name)
	local down_filename = string.format("%s_%s.down.sql", index_str, migration_name)

	local up_file_path = string.format("%s/%s", dir_migrations, up_filename)
	local down_file_path = string.format("%s/%s", dir_migrations, down_filename)

	write_file(up_file_path, "-- Up migration SQL goes here\n")

	if create_down_migration then
		write_file(down_file_path, "-- Down migration SQL goes here\n")
	end

	local log_message = string.format("Migration created successfully:\nUp: %s", up_file_path)
	if create_down_migration then
		log_message = log_message .. string.format("\nDown: %s", down_file_path)
	end
	log_info(log_message)
end

-- --- args parsing ----------------------------------------------------------

do
	local i = 1
	while i <= #arg do
		local a = arg[i]
		if a == "--down" or a == "-down" then
			create_down_migration = true
			table.remove(arg, i)
		else
			i = i + 1
		end
	end
end

migration_name = arg[1]

if not migration_name then
	usage()
	os.exit(1)
end

create_migration()
